apollo-federation/src/query_graph/graph_path.rs

```diff
@@ -2249,68 +2248,79 @@ impl OpGraphPath {
                 let build_err = || {
                     Err(FederationError::internal(format!(
-                        "{implem} is an object in {} but a {} in {}",
+                        "{implementation_type_pos} is an object in {} but a {} in {}",
                         tail.source, node.type_, node.source
                     )))
                 };
-                let QueryGraphNodeType::SchemaType(node_ty) = &node.type_ else {
+                let QueryGraphNodeType::SchemaType(node_type_pos) = &node.type_ else {
                     return build_err();
                 };
-                let node_ty = node_ty.get(schema)?;
-                let other_fields = match node_ty {
-                    ExtendedType::Object(obj) => &obj.fields,
-                    ExtendedType::Interface(int) => &int.fields,
-                    _ => return build_err(),
-                };
-                let Some(field) = other_fields.get(&itf.field_name) else {
+                let node_type_pos: ObjectOrInterfaceTypeDefinitionPosition =
+                    node_type_pos.clone().try_into()?;
+                let node_field_pos = node_type_pos.field(interface_field_pos.field_name.clone());
+                let Some(node_field) = node_field_pos.try_get(node_schema) else {
                     continue;
                 };
-                if !field.directives.has(&shareable_directive.name) {
+                if !node_field.directives.has(&node_shareable_directive.name) {
                     continue;
                 }
-                let field_ty = field.ty.inner_named_type();
-                if field_ty != base_ty_name
-                    || !(schema.get_object(field_ty).is_some()
-                        || schema.get_interface(field_ty).is_some())
-                {
+                let node_field_base_type_name = node_field.ty.inner_named_type();
+                if implementing_field_base_type_name != node_field_base_type_name {
                     // We have a genuine difference here, so we should explore type explosion.
                     return Ok(true);
                 }
-                let names: HashSet<_> = other_fields.keys().collect();
-                if !ty.fields.keys().all(|f| names.contains(&f)) {
-                    // Same, we have a genuine difference.
+                let node_field_base_type_pos =
+                    node_fed_schema.get_type(node_field_base_type_name.clone())?;
+                let Some(node_field_base_type_pos): Option<
+                    ObjectOrInterfaceTypeDefinitionPosition,
+                > = node_field_base_type_pos.try_into().ok() else {
+                    // Similar to above, we have a genuine difference.
+                    return Ok(true);
+                };
+
+                if !node_field_base_type_pos
+                    .fields(node_schema)?
+                    .all(|f| implementing_field_base_type_fields.contains(f.field_name()))
```

Comment:
```suggestion
                    .all(|f| implementing_field_base_type.fields.contains_key(f.field_name()))
```
I think collecting the keys separately is not neccessary here.

Reply:
It's indeed unnecessary, removed the extraneous hash set.